src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestHelpers/BaseRunner.cs

```diff
@@ -4,21 +4,133 @@
 using System;
 using System.Collections.Generic;
 using System.Text;
+using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.PowerFx.Core.Public.Values;
 
 namespace Microsoft.PowerFx.Core.Tests
 {
     // Base class for running a lightweght test. 
     public abstract class BaseRunner
-    {
+    {        
+        /// <summary>
+        /// Maximum time to run test - this catches potential hangs in the engine. 
+        /// Any test should easily run in under 1s. 
+        /// </summary>
+        public static TimeSpan Timeout = TimeSpan.FromSeconds(20);
+
         /// <summary>
         /// Runs a PowerFx test case, with optional setup.
         /// </summary>
         /// <param name="expr">PowerFx expression.</param>
         /// <param name="setupHandlerName">Optional name of a setup handler to run. Throws SetupHandlerNotImplemented if not found.</param>
         /// <returns>Result of evaluating Expr.</returns>
-        public abstract Task<FormulaValue> RunAsync(string expr, string setupHandlerName = null);
+        protected abstract Task<FormulaValue> RunAsyncInternal(string expr, string setupHandlerName = null);
+
+        /// <summary>
+        /// Returns (Pass,Fail,Skip) and a status message.
+        /// </summary>
+        /// <param name="test">test case to run.</param>
+        /// <returns>status from running.</returns>
+        public async Task<(TestResult, string)> RunAsync(TestCase testCase)
+        {
+            var result = TestResult.Fail;
+            string message = null;
+
+            var t = new Thread(() =>
```

Comment:
>var t = new Thread(() => [](http://example.com/codeflow?start=12&length=24)

XUnit doesn't support a timeout attribute. 
[https://github.com/xunit/xunit/issues/217](https://github.com/xunit/xunit/issues/217)

So offload to a thread on a timer.